store: collector for NetworkPolicy objects

milesbxf opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Exports information about NetworkPolicy objects

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #891

milesbxf commented:
Network policy metrics don't seem to be exposed in the e2e test rig. Anything obvious I've missed? Do I need to enable the collector somewhere?

brancz commented on internal/store/networkpolicy.go:
```diff
@@ -0,0 +1,125 @@
+						{
+							LabelKeys:   []string{},
+							LabelValues: []string{},
+							Value:       float64(len(n.Spec.Ingress)),
```
typically we don't do any aggregations in kube-state-metrics, why not expose a metric per rule?

LiliC replied:
I agree this can be done afterwards very easily.

brancz replied:
How do you mean afterwards?

LiliC replied:
Sorry, meant -> I agree with you @brancz as if needed the aggregation can be done later, and should not be done here, but we should expose raw metrics.

milesbxf replied:
I considered a raw metric per rule, but I was struggling to find a way to translate the `NetworkPolicy` rule spec into a sensible metric. For example, given this policy (from the docs at https://kubernetes.io/docs/concepts/services-networking/network-policies/): 
```
ingress:
  - from:
    - ipBlock:
        cidr: 172.17.0.0/16
        except:
        - 172.17.1.0/24
    - namespaceSelector:
        matchLabels:
          project: myproject
    - podSelector:
        matchLabels:
          role: frontend
    ports:
    - protocol: TCP
      port: 6379
  egress:
  - to:
    - ipBlock:
        cidr: 10.0.0.0/24
    ports:
    - protocol: TCP
      port: 5978
```

How would we sensibly represent this? The `from`/`to` blocks can have any number of sub-rules, with further nesting. I'm also not sure what the usefulness of this would be - it feels a bit like reimplementing the Kube API in metrics. Admittedly it'd be cool to be able to do things like count the number of times a particular label combination appears in NetworkPolicy rules, but I can't think how this could work 🤔 

Are there other resources we currently instrument in kube-state-metrics with complex nested structures like this?

milesbxf replied:
@tariq1890 @brancz @LiliC any thoughts on this? ☝️

brancz replied:
I don't think there is actually any precedence with this complexity. I think for this case I'd be ok with exposing counts, and come back to it should there be additional use cases. Can we make sure to mark those metrics as experimental?

LiliC commented on internal/store/networkpolicy.go:
```diff
@@ -0,0 +1,125 @@
+						{
+							LabelKeys:   []string{},
+							LabelValues: []string{},
+							Value:       float64(len(n.Spec.Egress)),
```
Same as above, we should adjust this.

LiliC commented on internal/store/networkpolicy_test.go:
```diff
@@ -0,0 +1,59 @@
+package store
```
Missing copyright header.

tariq1890 replied:
Good catch @LiliC! I am surprised we don't have a Copyright header checker added to our linting workflows.

tariq1890 commented on docs/README.md:
```diff
@@ -69,6 +69,7 @@ Per group of metrics there is one file for each metrics. See each file for speci
 - [Ingress Metrics](ingress-metrics.md)
 - [CertificateSigningRequest Metrics](certificatessigningrequest-metrics.md)
 - [VerticalPodAutoscaler Metrics](verticalpodautoscaler-metrics.md)
+- [NetworkPolicy Metrics](networkpolicy-metrics.md)
```
Something we haven't been enforcing well enough - lexicographical order. Let's place this above Node metrics.

milesbxf replied:
Would you like me to just sort the whole list alphabetically?

tariq1890 replied:
That would be great actually @milesbxf :)

tariq1890 commented on docs/networkpolicy-metrics.md:
```diff
@@ -0,0 +1,9 @@
+
+| Metric name                           | Metric type | Labels/tags                                 | Status       |
+| ------------------------------------- | ----------- | ------------------------------------------- | ------------ |
+| kube_networkpolicy_created            | Gauge       | `networkpolicy`=&lt;networkpolicy name&gt;  | EXPERIMENTAL |
```
These should be `STABLE` imo. They are all well-established fields in NetworkPolicy.

milesbxf replied:
👍 I wasn't sure what the convention was

LiliC replied:
@tariq1890 See my open issue :D https://github.com/kubernetes/kube-state-metrics/issues/892

@milesbxf sorry about this, we are still discussing, this we need to also document it better! :)

tariq1890 replied:
@LiliC Thanks for pointing that out. I think we can agree that these docs are fine as-is. So I'm resolving.

tariq1890 commented:
@milesbxf Can you add `networkpolicies` to the list of default collectors [here](https://github.com/kubernetes/kube-state-metrics/blob/c9d5deb4fa2305a4bf6216f8a6165f71e59869a3/pkg/options/collector.go#L28)? That should fix the e2e failures.

tariq1890 commented on internal/store/networkpolicy.go:
```diff
@@ -0,0 +1,125 @@
+		{
+			Name: "kube_networkpolicy_created",
+			Type: metric.Gauge,
+			Help: "Unix creation timestamp",
```
No strong opinions from here. None of the other metrics follow this convention @LiliC .

LiliC replied:
Yeah more like a nit for me :)

tariq1890 commented on internal/store/networkpolicy.go:
```diff
@@ -0,0 +1,125 @@
+)
+
+var (
+	descNetworkPolicyLabelsDefaultLabels = []string{"networkpolicy"}
```
NetworkPolicy is a namespaced kubernetes resource.
```suggestion
	descNetworkPolicyLabelsDefaultLabels = []string{"namespace", "networkpolicy"}
```

tariq1890 commented on internal/store/networkpolicy_test.go:
```diff
@@ -0,0 +1,59 @@
+	for i, c := range cases {
+		c.Func = metric.ComposeMetricGenFuncs(networkpolicyMetricFamilies)
+		if err := c.run(); err != nil {
+			t.Errorf("unexpected collecting result in %vth run:\n%s", i, err)
```
```suggestion
			t.Errorf("unexpected collecting result in %dth run:\n%v", i, err)
```

tariq1890 commented on internal/store/networkpolicy.go:
```diff
@@ -0,0 +1,125 @@
+
+		for _, m := range metricFamily.Metrics {
+			m.LabelKeys = append(descNetworkPolicyLabelsDefaultLabels, m.LabelKeys...)
+			m.LabelValues = append([]string{networkPolicy.Name}, m.LabelValues...)
```
```suggestion
			m.LabelValues = append([]string{networkPolicy.Namespace, networkPolicy.Name}, m.LabelValues...)
```

tariq1890 commented:
Thanks very much for this PR @milesbxf :)

tariq1890 commented on internal/store/networkpolicy_test.go:
```diff
@@ -0,0 +1,59 @@
+
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kube-state-metrics/pkg/metric"
```
This should have been caught by our linters. First-party import are to be separated from third-party ones.
```suggestion

	"k8s.io/kube-state-metrics/pkg/metric"
```

tariq1890 commented on internal/store/networkpolicy.go:
```diff
@@ -0,0 +1,125 @@
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kube-state-metrics/pkg/metric"
```
Wonder why the linters don't catch this. 🤔 
```suggestion

	"k8s.io/kube-state-metrics/pkg/metric"
```

tariq1890 commented:
@milesbxf Just checking in on this PR. Would you be able to rebase? I think we can just expose the counts here as mentioned by @brancz [here](https://github.com/kubernetes/kube-state-metrics/pull/893#discussion_r329600395)

milesbxf commented:
> @milesbxf Just checking in on this PR. Would you be able to rebase? I think we can just expose the counts here as mentioned by @brancz [here](https://github.com/kubernetes/kube-state-metrics/pull/893#discussion_r329600395)

I've rebased 👍 I noticed we moved to libsonnet for generating example YAMLs, so I've updated that too. I'll work through some of the other comments, and let you know once I think it's ready to finally review.

It looks like #914 only updated one of the generated examples and not
the libsonnet config itself, so I've added `admissionregistration.k8s.io`
and regenerated - do you think it's worth adding a CI check to catch this by making sure that `make examples` has been correctly run?

tariq1890 commented:
@milesbxf You might need to rebase again because of this PR #927 and add the tool dependencies in the new way. Sorry for the trouble :(

milesbxf commented:
> @milesbxf You might need to rebase again because of this PR #927 and add the tool dependencies in the new way. Sorry for the trouble :(

@tariq1890 done ✅

tariq1890 commented:
you would need to add the `gojsontoyaml` dep to `tools.go` so that it is registered in  the `go.mod` manifest as well.

milesbxf commented:
> you would need to add the `gojsontoyaml` dep to `tools.go` so that it is registered in the `go.mod` manifest as well.

Ah of course. Done as well 👍

milesbxf commented:
I've also run `go mod vendor` which should fix the failing CI check 🙈

tariq1890 commented:
@milesbxf I pulled your branch and I see that you haven't checked in the changes of `go mod vendor` completely.

There are some files in the `vendor` dir that would be deleted and changes in the `modules.txt`. Please rerun `go mod vendor` and make sure the resultant changes are git added.

milesbxf commented:
> @milesbxf I pulled your branch and I see that you haven't checked in the changes of `go mod vendor` completely.
> 
> There are some files in the `vendor` dir that would be deleted and changes in the `modules.txt`. Please rerun `go mod vendor` and make sure the resultant changes are git added.

Turns out I was running it with Go 1.12. I've rerun with 1.13 and pushed, let's hope this works 🙏

tariq1890 commented:
Awesome! Last request - let’s squash the commits :)

tariq1890 commented:
Oh no! It looks like merge conflicts need to be resolved :’(.

We will need to rebase once again.

tariq1890 commented:
/hold

tariq1890 commented:
/hold cancel

tariq1890 commented:
/lgtm

milesbxf commented:
@tariq1890 only just saw this - thanks so much for seeing this through and getting it merged! ❤️

tariq1890 commented:
You're most welcome! Thank you for your hard work with this :)
